src/main/java/com/amazon/ion/impl/bin/IonManagedWriter_1_1.kt

```diff
@@ -135,45 +145,299 @@ internal class IonManagedWriter_1_1(
         sid = newSymbols[text]
         if (sid != null) return sid
         // Add to the to-be-appended symbols
-        sid = priorMaxId + newSymbols.size + 1
+        sid = symbolTable.size + newSymbols.size + 1
         newSymbols[text] = sid
         return sid
     }
 
-    /** Writes a Local Symbol Table for the current encoding context, and updates the prior context. */
-    private fun writeSymbolTable() {
-        if (newSymbols.isEmpty()) return
-        val useSid = options.internEncodingDirectiveSymbols
+    private fun internMacro(macro: Macro): Int {
+        // Check the current macro table
+        var id = macroTable[macro]
+        if (id != null) return id
+        // Check the to-be-appended macros
+        id = newMacros[macro]
+        if (id != null) return id
+        // Add to the to-be-appended symbols
+        id = macrosById.size
+        macrosById.add(macro)
+        macroNames.add(null)
+        newMacros[macro] = id
+        return id
+    }
 
-        with(systemData) {
+    /** Converts a named macro reference to an address */
+    private fun MacroRef.ByName.intoId(): MacroRef.ById = MacroRef.ById(macroNames.indexOf(name))
 
-            if (useSid) writeAnnotations(SystemSymbols.ION_SYMBOL_TABLE_SID) else writeAnnotations(SystemSymbols.ION_SYMBOL_TABLE)
+    /** Converts a macro address to a macro name. If no name is found, returns the original address. */
+    private fun MacroRef.ById.intoNamed(): MacroRef = macroNames[id]?.let { MacroRef.ByName(it) } ?: this
 
-            systemData.stepInStruct(usingLengthPrefix = false)
-            if (canAppendToLocalSymbolTable) {
-                // LST Append
-                if (useSid) writeFieldName(SystemSymbols.IMPORTS_SID) else writeFieldName(SystemSymbols.IMPORTS)
-                if (useSid) writeSymbol(SystemSymbols.ION_SYMBOL_TABLE_SID) else writeSymbol(SystemSymbols.ION_SYMBOL_TABLE)
-            }
-            // ... and write the new symbols
-            if (useSid) writeFieldName(SystemSymbols.SYMBOLS_SID) else writeFieldName(SystemSymbols.SYMBOLS)
-            stepInList(usingLengthPrefix = false)
-            newSymbols.forEach { (text, _) -> writeString(text) }
-            stepOut()
-            stepOut()
+    private fun resetEncodingContext() {
```

Comment:
After this is called, how does the writer know that it needs to write a new encoding directive before the next value?

Reply:
The next time the user values are flushed, if `newSymbols` or `newMacros` is non-empty, then the managed writer will write an encoding directive before it flushes the user values. If both are empty, then we know that there are no SIDs or E-Expressions in the user values being flushed, and so no encoding directive is needed because it makes no difference.

Now that I think about it, there might be an edge case for Ion text if a system macro was being shadowed—it would end up writing incorrect data. This can be solved by making the managed writer emit an IVM if a value is written after `finish()` has been called.